### PR TITLE
common/BackTrace: extract demangle() out

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
ASan is not happy. i am going to look at the crimson side. but the reafactor in `common/BackTrace` is still valid.

File: src/common/BackTrace.cc [hunk 1]
```diff
@@ -76,61 +24,57 @@ void BackTrace::dump(Formatter *f) const
...
-	function = ret;
+  }
+  if (begin && end) {
+    std::string mangled(begin, end);
+    int status;
+    // only demangle a C++ mangled name
+    if (mangled.compare(0, 2, "_Z") == 0) {
+      // just a guess, template names will go much wider
+      size_t len = 1024;
+      char* buf = (char*)malloc(len);
```
Comment (rzarzynski) [inline src/common/BackTrace.cc:56]: Do we really need to allocate on our own?

File: src/common/BackTrace.cc [hunk 2]
```diff
@@ -76,61 +24,57 @@ void BackTrace::dump(Formatter *f) const
...
+      if (!buf) {
+        return {};
       }
-      else {
-	// demangling failed, just pretend it's a C function with no args
-	strncpy(function, foo, sz);
-	strncat(function, "()", sz);
-	function[sz-1] = 0;
+      if (char* demangled = abi::__cxa_demangle(mangled.c_str(), buf, &len, &status)) {
```
Comment (rzarzynski) [inline src/common/BackTrace.cc:60]: According to [a very old doc](https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a01696.html):

> output_buffer | A region of memory, allocated with malloc, of *length bytes, into which the demangled name is stored. If output_buffer is not long enough, it is expanded using realloc. output_buffer may instead be NULL; in that case, the demangled name is placed in a region of memory allocated with malloc.

File: src/common/BackTrace.cc (hunk 1 above)
Comment (tchaikov) [reply src/common/BackTrace.cc:56]: no. i will remove this. and think about a (premature) optimization to reuse this buffer in a separate commit.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-09-10_16:44:13-rados-wip-kefu-testing-2020-09-10-1633-distro-basic-smithi
- https://pulpito.ceph.com/kchai-2020-09-11_02:31:36-rados-wip-kefu-testing-2020-09-10-1633-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-09-11_06:48:10-rados-wip-kefu-testing-2020-09-10-1634-distro-basic-smithi/

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
@jdurgin @neha-ojha ping? could you help review the changes under `common`?

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2020-09-15_14:31:19-crimson-rados-wip-kefu-testing-2020-09-15-1950-distro-basic-smithi/

File: src/common/BackTrace.cc [hunk 3]
```diff
@@ -51,23 +51,17 @@ std::string BackTrace::demangle(const char* name)
     int status;
     // only demangle a C++ mangled name
     if (mangled.compare(0, 2, "_Z") == 0) {
-      // just a guess, template names will go much wider
-      size_t len = 1024;
-      char* buf = (char*)malloc(len);
```
Comment (rzarzynski) [inline src/common/BackTrace.cc]: Yeah, it's simpler and, likely, we don't need to care about optimizing this case at all. Simplicity won :+1: .

**Review by rzarzynski — APPROVED** [review]

**Pull request merged**